Update Dockerfile configuration

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This pull request represents a rework of this project's Docker configuration. It includes the following pull requests:
- #187
- #188
- #189
- #190
- #191
- #192
- #193
- #200
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

We have been striving to use configurations that are better at producing repeatable builds in downstream projects. This represents a backporting of a lot of that work to this skeleton so that all of our Docker projects are similarly configured.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

Automated tests pass.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Create a release.

mcdonnnj committed:
Use the full identifier for source Docker images

This helps ensure that when a Docker image is built the expacted source
image is used regardless of what repository is configured as the
default on the host system. It also makes our Dockerfiles more
seamlessly convertible to using the GitHub Container Registry or any
other Open Container Initiative (OCI) compatible registry.

mcdonnnj committed:
Use a specific version of Alpine Linux

Use the full tag that includes the Alpine Linux version to ensure the
pulled image is always the same.

mcdonnnj committed:
Merge pull request #187 from cisagov/improvement/use_full_image_source

Use the full path for source container images

mcdonnnj committed:
Install cisagov/skeleton-python-library directly

Instead of downloading the source archive, extracting it, and then
installing it with pip we instead just let pip directly install the
package.

mcdonnnj committed:
Remove unused OS package dependencies

Since we are now installing cisagov/skeleton-python-library directly
with pip we no longer need these OS packages.

mcdonnnj committed:
Change the secret message being checks in tests

Now that we are not overwriting the internal Python package file the
text we look for must match what is output by default. The Docker
Compose secret configuration is left in place to continue to serve as
an example and to be leveraged for a future update to
cisagov/skeleton-python-library that can provide similar functionality
to what was removed in this project.

mcdonnnj committed:
Remove package upgrading

We should not blindly upgrade all pre-installed packages. This can
create inconsistent build results due to changes in installed versions.

mcdonnnj committed:
Pin Python packages directly installed

Pin the versions of the pip, setuptools, and wheel packages that are
installed.

mcdonnnj committed:
Merge pull request #188 from cisagov/improvement/install_skeleton-python-library_directly

Install cisagov/skeleton-python-library directly with `pip`

mcdonnnj committed:
Move WORKDIR instruction

We can move this instruction to the end of the Dockerfile now that we
are no longer working with files in the Docker container when building.

mcdonnnj committed:
Merge pull request #189 from cisagov/improvement/make_builds_more_repeatable

Pin Python package versions and improve build repeatability

mcdonnnj committed:
Prefer calling pip as a module

Instead of relying on `pip3` being on the PATH we instead call the
module through the Python executable. This ensures that the `pip` being
used is in the same environment as the `python3` being used.

mcdonnnj committed:
Use a Python virtual environment in the Docker image

Using a virtual environment is a Python best practice. We also
consolidate all of the Python dependency installation steps into a
single RUN instruction. This ensures that Python setup is cached in one
layer and mirrors the logical organization of this being a single step.

mcdonnnj committed:
Explain `ln` options being used

Since we cannot use long options on Alpine Linux we should explain what
the short options we are using do. I also changed the order of options
so that they are in alphabetical order.

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

mcdonnnj committed:
Merge pull request #190 from cisagov/improvement/use_Python_venv

Use a Python virtual environment

mcdonnnj committed:
Add a pipenv configuration

This configuration includes a Pipfile configuration file and the
generated Pipfile.lock file that pins to specific versions for the
Python dependencies for this project. This will help us ensure
repeatable builds. The pipenv package is added as a developmental
requirement to support these files.

mcdonnnj committed:
Install Python dependencies using pipenv

Now that we have a pipenv configuration we will use it to install the
Python dependencies for the image. The `build` workflow is updated to
no longer pass the VERSION build argument in line with this change.

mcdonnnj committed:
Use a multi-stage Docker build

Switch to using a multi-stage build in the Dockerfile. This reduces
image size since pipenv and its dependencices are not needed in the
final image. It also ensures that the system Python environment is
unmodified.

mcdonnnj committed:
Install core Python packages into the system Python environment

Install the core Python packages (pip, setuptools, and wheel) into the
system Python environment before installing pipenv. This keeps things
consistent with our usual approach to Python environments.

mcdonnnj committed:
Fix outdated comment in the Dockerfile

The comment references a command that is no longer being run.

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

mcdonnnj committed:
Fix typo in Dockerfile comment

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj committed:
Merge pull request #191 from cisagov/improvement/pin_Python_configuration

Install Python dependencies with `pipenv`

mcdonnnj committed:
Bump setuptools from 69.1.0 to 69.1.1

mcdonnnj committed:
Bump Python from 3.12.0 to 3.12.2

mcdonnnj committed:
Bump Alpine Linux from 3.18 to 3.19

mcdonnnj committed:
Bump cisagov/skeleton-python-library from 0.0.1 to 0.2.0

Update the Dockerfile and testing to accommodate changes in the new
version.

mcdonnnj committed:
Update image tag information in the README

Change the tags used in the table to match the version of the project.
Previously "1.2.3" was used as an example version but there is no
reason not to use the real version of the image.

mcdonnnj committed:
Add instructions for managing Python dependencies

mcdonnnj committed:
Merge pull request #192 from cisagov/improvement/update_readme

Update the README

mcdonnnj committed:
Bump version from 0.0.1 to 0.2.0

mcdonnnj committed:
Merge pull request #193 from cisagov/improvement/update_dependencies

Update dependencies

dav3r commented on Dockerfile:
```diff
@@ -44,52 +87,25 @@ RUN addgroup --system --gid ${CISA_GID} ${CISA_GROUP} \
 
 ###
-# Dependencies
+# Copy in the Python virtual environment created in compile-stage, Sym-link the
```
Sorry for the nit-pick:
```suggestion
# Copy in the Python virtual environment created in compile-stage, symlink the
```

jsf9k replied:
That hyphen is distasteful to me.

jsf9k replied:
Thank you!

dav3r commented on Dockerfile:
```diff
@@ -44,52 +87,25 @@ RUN addgroup --system --gid ${CISA_GID} ${CISA_GROUP} \
-    && pip3 install --no-cache-dir --requirement requirements.txt \
-    && ln -snf /run/secrets/quote.txt src/example/data/secret.txt \
-    && rm sourcecode.tgz
+# Note that we sym-link the Python binary in the venv to the system-wide Python so that
```
Related nitpick:
```suggestion
# Note that we symlink the Python binary in the venv to the system-wide Python so that
```

dav3r reviewed (approved):
👍  after symlinkgate has been resolved.

mcdonnnj committed:
Correct usage of the term "symlink"

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

felddy commented on Dockerfile:
```diff
@@ -44,52 +87,25 @@ RUN addgroup --system --gid ${CISA_GID} ${CISA_GROUP} \
+# because the ln binary in Alpine Linux does not support long flags. The -f instructs
+# ln to remove the existing file and the -s instructs ln to create a symbolic link.
+###
+COPY --from=compile-stage --chown=${CISA_USER}:${CISA_GROUP} ${VIRTUAL_ENV} ${VIRTUAL_ENV}
```
TIL you can `chown` during a `COPY` in a Dockerfile.  Bravo.

felddy commented on Dockerfile:
```diff
@@ -1,8 +1,59 @@
+# cache.  This results in a smaller final image, at the cost of
+# slightly longer install times.
+###
+RUN python3 -m pip install --no-cache-dir --upgrade \
```
Since this is the `compile-stage` of the build, do we need to optimize for layer count.  i.e., does it still make sense to concatenate each command with `&&` versus having discreet `RUN`s.  

I ask because I think you will get a more localized error statement when something fails when the command is decomposed.  There may be some other gains to be had as well with caching, and parallelization.

mcdonnnj replied:
I don't know that there's much to be gained from breaking this `RUN` statement up. I think any errors will be fairly obvious because there are only four parts and only the first and last are similar in what's being done. I'm also not convinced there's much to be gained in terms of caching or especially parallelization. Each part _must_ be done in sequence to have an expected result.

felddy commented on Dockerfile:
```diff
@@ -44,52 +87,25 @@ RUN addgroup --system --gid ${CISA_GID} ${CISA_GROUP} \
-    && pip3 install --no-cache-dir --requirement requirements.txt \
-    && ln -snf /run/secrets/quote.txt src/example/data/secret.txt \
-    && rm sourcecode.tgz
+# Note that we symlink the Python binary in the venv to the system-wide Python so that
```
Should this comment be moved down to after the `COPY` but before the  `RUN ln ...` line?

mcdonnnj replied:
Since the header for that comment block describes everything being done in the following three commands I think it's fine as-is. That's just an information note about a "why".

felddy reviewed (approved):
Fully charged!  🔋 
You deserve an award for this modernization!  🏆 
I had a few comments.

mcdonnnj committed:
Bump python from 3.12.2-alpine3.19 to 3.13.1-alpine3.20

The version of Python listed in the Pipfile is updated to match the new
Docker image tag.

mcdonnnj committed:
Bump Python packages installed in the Dockerfile

- pip from 24.0 to 24.3.1
- pipenv from 2023.12.1 to 2024.4.0
- setuptools from 69.1.1 to 75.6.0
- wheel from 0.42.0 to 0.45.1

mcdonnnj committed:
Update Python virtual environment dependencies

Update the dependencies installed in the Python virtual environment by
running `pipenv lock` in the `src/` directory.

mcdonnnj committed:
Merge pull request #200 from cisagov/improvement/update_dependencies

Update image dependencies

mcdonnnj committed:
Add a missing Oxford comma to a comment in the Dockerfile

jsf9k commented on Dockerfile:
```diff
@@ -1,8 +1,59 @@
-ARG VERSION=unspecified
+# Official Docker images are in the form library/<app> while non-official
+# images are in the form <user>/<app>.
+FROM docker.io/library/python:3.13.1-alpine3.20 as compile-stage
```
Are the warnings here addressed elsewhere?  If not then we may as well address them here.

mcdonnnj replied:
Ah, right, the casing warnings that have cropped up more recently. Please see [daeb027](https://github.com/cisagov/skeleton-docker/pull/195/commits/daeb027aa83e62f8723e974d459cd3ae528925ae).

mcdonnnj committed:
Ensure FROM/AS keywords use the same casing

This resolves the following warning from Docker when building the
image:
FromAsCasing: 'as' and 'FROM' keywords' casing do not match

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>
